src/composables/nav.ts

```diff
@@ -21,3 +29,204 @@ export interface ActionItem {
+      { tab: tab },
+      document.title,
+      `${window.location.pathname}${queryString}`
+    )
+  })
+
+  return {
+    activeTab,
+    tabs,
+  }
+}
+
+/**
+ * URLParamValue is the set of value types that are
+ * currently spec'd to be parsed into a query string.
+ */
+export type URLParamValue =
+  | string
+  | number
+  | boolean
+  | null
+  | undefined
+  | string[]
+  | number[]
+
+/**
+ * URLParams generic serves as a type constraint ensuring that
+ * the search params record used only contains fields that are
+ * currently spec'd to be parsed into a query string. See: URLParamValue
+ *
+ * Example:
+ *
+ * interface SearchParams {
+ *    // All Good!
+ *    q: string
+ *    count: number
+ *    isActive: boolean
+ *    terms: string[]
+ *    bucketOfNumbers: number[]
+ *
+ *    // Nope!
+ *    nested: {
+ *        qTwo: string
+ *    }
+ * }
+ */
+export type URLParams<T> = {
+  [P in keyof T]: T[P] extends URLParamValue ? T[P] : never
+}
+
+/**
+ * useUrlSearchParams accepts an initial set of search query params
+ * in the shape of Record<string, URLParamValue> and tracks the values
+ * of each param on window.location.search.
+ *
+ * This is a one-way sync where non-nullish, non-falsey values are popped
+ * onto window.location.search as they are mutated.  The initial values are not read
+ * from an existing query string on window.location.search as each values
+ * type cannot be inferred safely.
+ *
+ * This composable ignores query params that may already exist on window.location.search
+ * leaving the intact as much as possible, since it's not possible to determine where they
+ * came from or what function they may serve.
+ *
+ * The return value is a ref of the initial set of params.
+ *
+ * Example Usage:
+ * interface SearchParams {
+ *    q: string
+ *    isActive: boolean
+ * }
+ *
+ * // NOTE: initial params values should come from the
+ * // server in the form of a page prop. Though some
+ * // exceptions may exist.
+ * const props = defineProps<{
+ *    initialParams: SearchParams
+ * }>
+ *
+ * const searchParams = useUrlSearchParams(props.initialParam)
+ *
+ * <BaseInput v-model="searchParams.q" label="Search" type="search" />
+ * <Checkbox v-model="searchParams.isActive" label="Is Active" />
+ *
+ *  When q = "jimothy bobbitz" and isActive = false
+ * /path?q=jimothy+bobbitz
+ *
+ * When q = "jimothy bobbitz" and isActive = true
+ * /path?q=jimothy+bobbitz&isActive=true
+ *
+ * * When q = "" and isActive = false
+ * /path
+ *
+ * @param initial any Record<string, URLParamValue> like interface
+ * @return Ref<T>
+ */
+export const useUrlSearchParams = <T>(initial: URLParams<T>) => {
+  const searchParams = ref(initial)
+
+  watch(
+    searchParams,
+    (p) => {
+      const params = new URLSearchParams(window.location.search)
+
+      for (const key in p) {
+        const val = p[key]
+        params.delete(key)
+
+        if (val == "" || val == false || val == null || val == undefined) {
```

Comment:
Probably just not grokking the overall context, but I'm confused why false values are thrown away. I imagine `showEmails: false` and `showEmails: true` and an empty value trigger 3 separate queries. Getting rid of the key would block necessary behavior in my mind.

Reply:
> Getting rid of the key would block necessary behavior in my mind.

I probably overlooked some use cases in order to keep this first iteration simple (i.e. no configuration options).  On community search, this is what we're doing via the `removeNullishValues` and `removeFalsyValues` options in the `VueUse` version of [useUrlSearchParams](https://vueuse.org/core/useUrlSearchParams/?foo=bar&vueuse=awesome#type-declarations).

By itself,  `useUrlSearchParams` won't trigger queries as it's a one-way binding between some object with key/value pairs and `window.location.search` where `window.location.search` is mutated as the object is mutated, so generally, it's just reacting to input controls.  Queries are fired off based on mutations of the same object, but some client-side method and Axios ultimately determine the shape of that payload.

Axios, currently drops `nullish` values from payloads (as we have it configured), so I expect, we'll never need to record nullish values via `useUrlSearchParams`, primarily because our input controls which can be initialized with a null value, don't _typically_ allow the user to return to a null value once they've interacted with it (think radios and checkboxes) - though exceptions exist.  I believe a `SearchParams` like struct could also end up with a literal "null" string value if not carefully considered.

However, I can see how a `showEmails: null | true | false` might evaluate to 3 different results in a query, so if we supported specifically "false" values we should be able to cover those cases, I'm just not clear on how common that really is or whether we should do it by default at the moment.

There's also a little bit of nuance here around whether `useUrlSearchParams` should immediately evaluate initialParams or not.  Currently, it does not, since I wanted to avoid mutating browser history on initialization.

Reply:
@profsmallpine let me know if you checked out the side effect of the change on the community search page in [#1091](https://github.com/xy-planning-network/college-try/pull/1091).  It does set all params on first mutation, i.e.

`/community/profiles?q=test+search&sort=&sortDir=&ethnicity=&feeFlat=false&feeHourly=false&feeMonthly=false&feeQuarterly=false&feeAUM=false&feeAdditional=false&genderIdentity=&hasEmploymentInterest=false&hasHiringInterest=false&hasPurchaseInterest=false&hasSaleInterest=false&hasMergeInterest=false&hasMenteeInterest=false&hasMentorInterest=false&persona=&teamSize=&state=&location=`

It should be fine, just a notable difference when there are many search params (which does seem like an outlier).

If you have a place in archive you want to pull this into before I merge and publish to NPM let me know.  Latest is available via `npm install @xy-planning-network/trees@dev`.

Reply:
Yeah that gets busy. Do we want to continue dropping null values and just set true/false? I was most concerned about the need to have false values in the param, less so with null and that keeps the url more manageable. Maybe it doesn't matter much as users don't give it a ton of attention.

Reply:
> Do we want to continue dropping null values 

We are dropping them in the current implementation - we have to to support the parser/decoder on the server side.  False is being set in community search since most of the checkbox like filters are zero-valued to `false`.

> Maybe it doesn't matter much as users don't give it a ton of attention.

This is my expectation, but there is something that doesn't feel right about it.  Maybe we're just over thinking it.

I can't easily determine what key's have/haven't mutated with a watcher on an object. Typically, the callback receives `newVal`, `oldVal` arguments, but in the case of an object like this, Vue doesn't hold the `oldVal`.  Since there are array's in the mix I also can't just use a simple shallow copy, though I could do a deep copy, and in theory have a reference to work with.

Reply:
or, we can relent and just make it an option `{removeFalsey: bool}`

Reply:
Did some testing on portal, this strikes a good balance 👍 🚀